prov/efa/src/efa_msg.c

```diff
@@ -0,0 +1,408 @@
+
+	err = ibv_post_recv(qp->ibv_qp, &base_ep->efa_recv_wr_vec[0].wr, &bad_wr);
+	if (OFI_UNLIKELY(err)) {
+		/* On failure, ibv_post_recv() return positive errno.
+		 * Meanwhile, this function return a negative errno.
+		 * So, we do the conversion here.
+		 */
+		err = (err == ENOMEM) ? -FI_EAGAIN : -err;
+	}
+
+	base_ep->efa_recv_wr_index = 0;
+
+	return err;
+
+out_err:
+	if (wr_index > 0) {
+		post_recv_err = ibv_post_recv(qp->ibv_qp, &base_ep->efa_recv_wr_vec[0].wr, &bad_wr);
+		if (post_recv_err) {
+			EFA_WARN(FI_LOG_EP_DATA,
+				 "Encountered error %ld when ibv_post_recv on error handling path\n",
+				 post_recv_err);
+		}
+	}
+
+	return err;
+}
+
+static ssize_t efa_ep_recvmsg(struct fid_ep *ep_fid, const struct fi_msg *msg, uint64_t flags)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+
+	return efa_post_recv(base_ep, msg, flags);
+}
+
+static ssize_t efa_ep_recv(struct fid_ep *ep_fid, void *buf, size_t len,
+			   void *desc, fi_addr_t src_addr, void *context)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+	struct iovec iov;
+	struct fi_msg msg;
+
+	EFA_SETUP_IOV(iov, buf, len);
+	EFA_SETUP_MSG(msg, &iov, &desc, 1, src_addr, context, 0);
+
+	return efa_post_recv(base_ep, &msg, 0);
+}
+
+static ssize_t efa_ep_recvv(struct fid_ep *ep_fid, const struct iovec *iov, void **desc,
+			    size_t count, fi_addr_t src_addr, void *context)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+	struct fi_msg msg;
+
+	EFA_SETUP_MSG(msg, iov, desc, count, src_addr, context, 0);
+
+	return efa_post_recv(base_ep, &msg, 0);
+}
+
+static inline ssize_t efa_post_send_validate(struct efa_base_ep *base_ep, const struct fi_msg *msg,
+				      struct efa_conn *conn, uint64_t flags, size_t *len)
+{
+	if (OFI_UNLIKELY(msg->iov_count > base_ep->info->tx_attr->iov_limit)) {
+		EFA_WARN(FI_LOG_EP_DATA, "requested sge[%zu] is greater than max supported[%zu]!\n",
+			 msg->iov_count, base_ep->info->tx_attr->iov_limit);
+		return -FI_EINVAL;
+	}
+
+	if (OFI_UNLIKELY(msg->msg_iov[0].iov_len <
+			 base_ep->info->ep_attr->msg_prefix_size)) {
+		EFA_WARN(FI_LOG_EP_DATA, "prefix not present on first iov, iov_len[%zu]\n",
+			 msg->msg_iov[0].iov_len);
+		return -EINVAL;
+	}
+
+	*len = ofi_total_iov_len(msg->msg_iov, msg->iov_count) - base_ep->info->ep_attr->msg_prefix_size;
+	if (OFI_UNLIKELY(*len > base_ep->info->ep_attr->max_msg_size)) {
+		EFA_WARN(FI_LOG_EP_DATA, "requested size[%zu] is greater than max[%zu]!\n",
+			 *len, base_ep->info->ep_attr->max_msg_size);
+		return -FI_EINVAL;
+	}
+
+	return 0;
+}
+
+static inline bool efa_msg_has_hmem_mr(const struct fi_msg *msg)
+{
+	/* the device only support send up 2 iov, so iov_count cannot be > 2 */
+	assert(msg->iov_count == 1 || msg->iov_count == 2);
+	/* first iov is always on host memory, because it must contain packet header */
+	assert(!efa_mr_is_hmem(msg->desc[0]));
```

Comment:
This was wrong, we should shift the iov by prefix size and only check the payload